assets/src/asset.rs

```diff
@@ -0,0 +1,164 @@
+/// This data is then generated by the `Format` trait.
+pub trait Asset
+    where Self: Sized
+{
+    type Context;
+    type Data;
+    type Error: Error;
+
+    /// A small keyword for which category this asset belongs to.
+    ///
+    /// ## Examples
+    ///
+    /// * `"mesh"` for `Mesh`
+    /// * `"data"` for `Level`
+    ///
+    /// The storage may use this information, to e.g. search the identically-named
+    /// subfolder.
+    fn category() -> &'static str;
+
+    /// Provides the conversion from the data format to the actual asset.
+    fn from_data(data: Self::Data, context: &Self::Context) -> Result<Self, Self::Error>;
+
+    /// Notifies about an asset load. This is can be used to cache the asset.
+    /// To return a cached asset, see the `cached` function.
+    fn asset_loaded(_context: &Self::Context, _spec: AssetSpec, _asset: &Self) {}
+
+    /// Returns `Some` cached value if possible, otherwise `None`.
+    ///
+    /// For a basic implementation of a cache, please take a look at the `Cache` type.
+    fn cached(_context: &Self::Context, _spec: &AssetSpec) -> Option<Self> {
+        None
+    }
+
+    /// Gives a hint that several assets may have been released recently.
+    ///
+    /// This is useful if your assets are reference counted, because you are
+    /// now able to remove unique assets from the cache, leaving the shared
+    /// ones there.
+    fn clear(_context: &Self::Context) {}
+
+    /// Request for clearing the whole cache.
+    fn clear_all(_context: &Self::Context) {}
+}
+
+/// A specifier for an asset, uniquely identifying it by
+///
+/// * the extension (the format it was provided in)
+/// * it's name
+/// * the storage it was loaded from
+#[derive(Clone, Debug, Eq, Hash, Ord, PartialEq, PartialOrd)]
+pub struct AssetSpec {
+    /// The extension of this asset
+    pub ext: &'static str,
+    /// The name of this asset.
+    pub name: String,
+    /// The storage id, uniquely identifying the storage it's been loaded from.
+    pub store: StoreId,
+}
+
+impl AssetSpec {
+    /// Creates a new asset specifier from the given parameters.
+    pub fn new(name: String, ext: &'static str, store: StoreId) -> Self {
+        AssetSpec {
+            ext,
+            name,
+            store,
+        }
+    }
+}
+
+/// A basic implementation for a cache. This might be useful as the `Context` of
+/// an `Asset`, so that the same asset doesn't get imported twice.
+///
+/// Because contexts have to be immutable, a `RwLock` is used. Therefore, all
+/// operations are blocking (but shouldn't block for a long time).
+pub struct Cache<T> {
+    map: RwLock<FnvHashMap<AssetSpec, T>>,
+}
+
+impl<T> Cache<T>
+    where T: Clone
+{
+    /// Creates a new `Cache` and initializes it with the default values.
+    pub fn new() -> Self {
+        Default::default()
+    }
+
+    /// Inserts an asset, locking the internal `RwLock` to get write access to the hash map.
+    pub fn insert(&self, spec: AssetSpec, asset: T) {
```

Comment:
If a value exists for the given key in a HashMap the insert operation will return the value it replaced.  It might be helpful to have this function do the same, so that users of this function can identify when they did a redundant insert, and modify their code to not do those redundant inserts.  It's certainly not necessary, it's just helpful for debugging.